randovania/layout/generator_parameters.py

```diff
@@ -0,0 +1,113 @@
+import json
+from dataclasses import dataclass
+from typing import Dict, Iterator, Tuple
+
+from randovania.bitpacking import bitpacking
+from randovania.bitpacking.bitpacking import BitPackValue, BitPackDecoder
+from randovania.games import default_data
+from randovania.games.game import RandovaniaGame
+from randovania.interface_common.preset_manager import PresetManager
+from randovania.layout.preset import Preset
+
+_PERMALINK_MAX_SEED = 2 ** 31
+_PERMALINK_PLAYER_COUNT_LIMITS = (2, 256)
+
+
+def _game_db_hash(game: RandovaniaGame) -> int:
+    data = default_data.read_json_then_binary(game)[1]
+    return bitpacking.single_byte_hash(json.dumps(data, separators=(',', ':')).encode("UTF-8"))
+
+
+@dataclass(frozen=True)
+class GeneratorParameters(BitPackValue):
+    seed_number: int
+    spoiler: bool
+    presets: Dict[int, Preset]
+
+    def __post_init__(self):
+        if self.seed_number is None:
+            raise ValueError("Missing seed number")
+        if not (0 <= self.seed_number < _PERMALINK_MAX_SEED):
+            raise ValueError("Invalid seed number: {}".format(self.seed_number))
+        object.__setattr__(self, "__cached_as_bytes", None)
+
+    def bit_pack_encode(self, metadata) -> Iterator[Tuple[int, int]]:
+        yield self.seed_number, _PERMALINK_MAX_SEED
+        yield from bitpacking.encode_bool(self.spoiler)
+        yield from bitpacking.encode_int_with_limits(self.player_count, _PERMALINK_PLAYER_COUNT_LIMITS)
+
+        manager = PresetManager(None)
+
+        previous_unique_presets = []
+        for preset in self.presets.values():
+            already_encoded_preset = preset in previous_unique_presets
+            yield from bitpacking.encode_bool(already_encoded_preset)
+            if already_encoded_preset:
+                yield from bitpacking.pack_array_element(preset, previous_unique_presets)
+            else:
+                previous_unique_presets.append(preset)
+                yield from preset.bit_pack_encode({"manager": manager})
+
+        games = []
+        for preset in self.presets.values():
+            if preset.game not in games:
+                games.append(preset.game)
+                yield _game_db_hash(preset.game), 256
```

Comment:
set comprehension might be cleaner:

```py
games = {preset.game for preset in self.presets.values()}
for game in games:
    yield _game_db_hash(game), 256
```

Reply:
not deterministic

Reply:
could always just sort it, but yeah makes sense